chalk-solve/src/clauses/clause_visitor.rs

```diff
@@ -0,0 +1,112 @@
+    in_clauses: &Vec<ProgramClause>,
+    out: &mut FxHashSet<ProgramClause>,
+) {
+    let mut visitor = ClauseVisitor::new(db, out);
+    for clause in in_clauses {
+        visitor.visit_program_clause(&clause);
+    }
+}
+
+struct ClauseVisitor<'db, 'set> {
+    db: &'db dyn RustIrDatabase,
+    round: &'set mut FxHashSet<ProgramClause>,
+}
+
+impl<'db, 'set> ClauseVisitor<'db, 'set> {
+    fn new(db: &'db dyn RustIrDatabase, round: &'set mut FxHashSet<ProgramClause>) -> Self {
+        ClauseVisitor { db, round }
+    }
+
+    fn visit_projection_ty(&mut self, projection_ty: &ProjectionTy) {
+        let mut clauses = vec![];
+        self.db
+            .associated_ty_data(projection_ty.associated_ty_id)
+            .to_program_clauses(self.db, &mut clauses);
+        self.round.extend(clauses);
+    }
+
+    fn visit_ty(&mut self, ty: &Ty) {
+        let mut clauses = vec![];
+        match ty {
+            Ty::Apply(application_ty) => match application_ty.name {
+                TypeName::TypeKindId(type_kind_id) => {
+                    match_type_kind(self.db, type_kind_id, &mut clauses)
+                }
+                TypeName::Placeholder(_) => (),
+                TypeName::AssociatedType(type_id) => {
+                    self.db
+                        .associated_ty_data(type_id)
+                        .to_program_clauses(self.db, &mut clauses);
+                }
+            },
+            Ty::Projection(projection_ty) => {
+                self.visit_projection_ty(projection_ty);
+            }
+            Ty::UnselectedProjection(_) | Ty::ForAll(_) | Ty::BoundVar(_) | Ty::InferenceVar(_) => {
+                ()
+            }
+        }
+        self.round.extend(clauses);
+    }
+
+    fn visit_from_env(&mut self, from_env: &FromEnv) {
+        match from_env {
+            FromEnv::Trait(trait_ref) => {
+                let mut clauses = vec![];
+                let trait_datum = self.db.trait_datum(trait_ref.trait_id);
+
+                trait_datum.to_program_clauses(self.db, &mut clauses);
+
+                // If we know that `T: Iterator`, then we also know
+                // things about `<T as Iterator>::Item`, so push those
+                // implied bounds too:
+                for &associated_ty_id in &trait_datum.binders.value.associated_ty_ids {
+                    self.db
+                        .associated_ty_data(associated_ty_id)
+                        .to_program_clauses(self.db, &mut clauses);
+                }
+
+                self.round.extend(clauses);
+            }
+            FromEnv::Ty(ty) => self.visit_ty(ty),
+        }
+    }
+
+    fn visit_domain_goal(&mut self, domain_goal: &DomainGoal) {
+        match domain_goal {
+            DomainGoal::FromEnv(from_env) => self.visit_from_env(from_env),
+            DomainGoal::Holds(WhereClause::ProjectionEq(ProjectionEq { projection, ty: _ })) => {
+                self.visit_projection_ty(projection)
```

Comment:
I’m not sure this is useful; you would not have any useful implied bounds if you just have e.g. `where ProjectionEq(<T as Iterator>::Item = HashSet<K>)`

Reply:
Maybe it's not needed. I'm experimenting with removing it.